[mlir][spirv] Add GpuToLLVM cconv suited to Vulkan, migrate last tests

andfau-amd opened the pull request:
This commit is a follow-up to 99a562b3cb17e89273ba0fe77129f2fb17a19381, which migrated some of the mlir-vulkan-runner tests to mlir-cpu-runner using a new pipeline and set of wrappers. That commit could not migrate all the tests, because the existing calling conventions/ABIs for kernel arguments generated by GPUToLLVMConversionPass were not a good fit for the Vulkan runtime. This commit fixes this and migrates the remaining tests. With this commit, mlir-vulkan-runner and many related components are now unused, and they will be removed in a later commit (see #73457).

The old calling conventions require both the caller (host LLVM code) and callee (device code) to have compile-time knowledge of the precise argument types. This works for CUDA, ROCm and SYCL, where there is a C-like calling convention agreed between the host and device code, and the runtime passes through arguments as raw data without comprehension. For Vulkan, however, the interface declared by the shader/kernel is in a more abstract form, so the device code has indirect access to the argument data, and the runtime must process the arguments to set up and bind appropriately-sized buffer descriptors.

This commit introduces a new calling convention option to meet the Vulkan runtime's needs. It lowers memref arguments to {void*, size_t} pairs, which can be trivially interpreted by the runtime without it needing to know the original argument types. Unlike the stopgap measure in the previous commit, this system can support memrefs of various ranks and element types, which unblocked migrating the remaining tests.

andfau-amd commented on mlir/lib/Conversion/GPUCommon/GPUToLLVMConversion.cpp:
```diff
@@ -970,33 +971,56 @@ LogicalResult LegalizeLaunchFuncOpPattern::matchAndRewrite(
             launchOp, "Operand to launch op is not a memref.");
       }
+
+      if (!memrefTy.hasStaticShape() ||
+          !memrefTy.getElementType().isIntOrFloat()) {
+        return rewriter.notifyMatchFailure(
+            launchOp, "Operand to launch op is not a memref with a static "
+                      "shape and an integer or float element type.");
+      }
+
+      unsigned bitwidth = memrefTy.getElementTypeBitWidth();
+      if (bitwidth % 8 != 0) {
+        return rewriter.notifyMatchFailure(
+            launchOp, "Operand to launch op is not a memref with a "
+                      "byte-aligned element type.");
+      }
```
I feel like I can't be the first person to have wanted to do this, and I'm not sure I'm checking all the necessary conditions. Is there some handy function for this that I'm missing?

kuhar replied:
I don't recall seeing a helper for this either. Most code uses `llvm::ceilDiv` for storage size though instead of rejecting element sizes that are not multiples of 8 bits.

andfau-amd replied:
Is there some rule somewhere that guarantees elements smaller than 8 bits get padded to 8?

I doubt the Vulkan target can actually do anything useful with non-8-bit types though, so maybe this simple check is fine.

andfau-amd commented on mlir/lib/Conversion/GPUCommon/GPUToLLVMConversion.cpp:
```diff
@@ -970,33 +971,56 @@ LogicalResult LegalizeLaunchFuncOpPattern::matchAndRewrite(
+      llvmArgumentsWithSizes.push_back(llvmArg); // presumably bare pointer
+      llvmArgumentsWithSizes.push_back(sizeArg);
     }
+    llvmArguments = std::move(llvmArgumentsWithSizes);
```
This is pretty ugly but I don't have a better idea. Inserting into an existing list would be ugly and slow (in a very theoretical sense).

andfau-amd replied:
I decided I disliked this enough to change it to having two arrays on the stack and a conditional selection. The advantage of this setup is there's no move from a `SmallVector<T, 8>` into a `SmallVector<T, 4>`, which had been making me wince a bit.

Hardcode84 commented on mlir/include/mlir/Conversion/GPUCommon/GPUCommonPass.h:
```diff
@@ -62,10 +62,10 @@ struct FunctionCallBuilder {
-                                         bool typeCheckKernelArgs = false);
+void populateGpuToLLVMConversionPatterns(
+    LLVMTypeConverter &converter, RewritePatternSet &patterns,
+    bool kernelBarePtrCallConv = false,
```
Maybe better to make it enum instead of 2 bools?

andfau-amd replied:
(Specifically the `Option` defined in `Passes.td`. I know that it's somehow leveraging the `cl` infrastructure so enums "should" be possible but I haven't managed to figure it out from staring at the headers and tablegen definitions, there's too many layers.)

Hardcode84 replied:
Here is the example https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Conversion/Passes.td#L1145, but we can do it in separate PR

Hardcode84 commented on mlir/lib/Conversion/GPUCommon/GPUToLLVMConversion.cpp:
```diff
@@ -970,33 +971,56 @@ LogicalResult LegalizeLaunchFuncOpPattern::matchAndRewrite(
+          "Cannot add sizes to arguments with one-to-many LLVM IR expansion.");
+    }
+
+    SmallVector<Value, 8> llvmArgumentsWithSizes;
```
nit: you can drop `8` in `SmallVector` in most cases.

andfau-amd replied:
In the rearranged code I have kept the `8` but it's for the sake of clarity (one array has two times the size of the other). Does that seem okay?

andfau-amd replied:
Oops, the conditional move means I can't have one be 4 and the other be 8, they have to both be 8. Well, I want to use this explicit size because it is the right one for the actual use-case.

kuhar commented on mlir/lib/Conversion/GPUCommon/GPUToLLVMConversion.cpp:
```diff
@@ -970,33 +971,56 @@ LogicalResult LegalizeLaunchFuncOpPattern::matchAndRewrite(
+      }
+
+      uint64_t staticSize =
+          uint64_t(bitwidth / 8) * uint64_t(memrefTy.getNumElements());
```
prefer `static_cast`.

kuhar commented on mlir/lib/Conversion/GPUCommon/GPUToLLVMConversion.cpp:
```diff
@@ -970,33 +971,56 @@ LogicalResult LegalizeLaunchFuncOpPattern::matchAndRewrite(
+
+      Value sizeArg = rewriter.create<LLVM::ConstantOp>(
+          loc, getIndexType(), rewriter.getIndexAttr(staticSize));
+      llvmArgumentsWithSizes.push_back(llvmArg); // presumably bare pointer
```
```suggestion
      llvmArgumentsWithSizes.push_back(llvmArg); // Presumably a bare pointer.
```

see https://llvm.org/docs/CodingStandards.html#commenting

kuhar commented on mlir/test/lib/Pass/TestVulkanRunnerPipeline.cpp:
```diff
@@ -68,12 +68,11 @@ void buildTestVulkanRunnerPipeline(OpPassManager &passManager,
         LLVM::createRequestCWrappersPass());
-    // vulkan-runtime-wrappers.cpp uses the non-bare-pointer calling convention,
-    // and the type check is needed to prevent accidental ABI mismatches.
+    // These calling convention options match vulkan-runtime-wrappers.cpp
```
```suggestion
    // These calling convention options match vulkan-runtime-wrappers.cpp.
```

kuhar commented on mlir/tools/mlir-vulkan-runner/vulkan-runtime-wrappers.cpp:
```diff
@@ -169,26 +169,22 @@ mgpuLaunchKernel(void *vkKernel, size_t gridX, size_t gridY, size_t gridZ,
+    size_t memrefBufferSize = *static_cast<size_t *>(params[i + 1]);
+    VulkanHostMemoryBuffer memBuffer{memrefBufferBasePtr,
+                                     static_cast<uint32_t>(memrefBufferSize)};
+    reinterpret_cast<VulkanRuntimeManager *>(vkRuntimeManager)
```
```suggestion
    manager
```

?

andfau-amd replied:
Oops yeah I missed that this was redundant when copying from `bindMemref`, thanks.

kuhar reviewed (commented):
Looks good overall, just some nits

andfau-amd committed:
[mlir][spirv] Add GpuToLLVM cconv suited to Vulkan, migrate last tests

This commit is a follow-up to 99a562b3cb17e89273ba0fe77129f2fb17a19381,
which migrated some of the mlir-vulkan-runner tests to mlir-cpu-runner
using a new pipeline and set of wrappers. That commit could not migrate
all the tests, because the existing calling conventions/ABIs for kernel
arguments generated by GPUToLLVMConversionPass were not a good fit for
the Vulkan runtime. This commit fixes this and migrates the remaining
tests. With this commit, mlir-vulkan-runner and many related components
are now unused, and they will be removed in a later commit (see #73457).

The old calling conventions require both the caller (host LLVM code)
and callee (device code) to have compile-time knowledge of the precise
argument types. This works for CUDA, ROCm and SYCL, where there is a
C-like calling convention agreed between the host and device code, and
the runtime passes through arguments as raw data without comprehension.
For Vulkan, however, the interface declared by the shader/kernel is in a
more abstract form, so the device code has indirect access to the
argument data, and the runtime must process the arguments to set up and
bind appropriately-sized buffer descriptors.

This commit introduces a new calling convention option to meet the
Vulkan runtime's needs. It lowers memref arguments to {void*, size_t}
pairs, which can be trivially interpreted by the runtime without it
needing to know the original argument types. Unlike the stopgap measure
in the previous commit, this system can support memrefs of various ranks
and element types, which unblocked migrating the remaining tests.

kuhar reviewed (approved):
LGTM

Hardcode84 commented on mlir/lib/Conversion/GPUCommon/GPUToLLVMConversion.cpp:
```diff
@@ -970,33 +971,55 @@ LogicalResult LegalizeLaunchFuncOpPattern::matchAndRewrite(
+  SmallVector<Value, 8> llvmArguments = getTypeConverter()->promoteOperands(
+      loc, origArguments, adaptor.getKernelOperands(), rewriter,
+      /*useBarePtrCallConv=*/kernelBarePtrCallConv);
+  SmallVector<Value, 8> llvmArgumentsWithSizes;
```
nit: don't need to explicitly specify `8` unless this is some special case (`0`/`1`)
